File: microvm/foreground.py
"""Local animated view consuming the guest's display journal over shell ingress."""
import json, os, re, select, subprocess, sys, tempfile, time
from pathlib import Path

def recent_turns(session):
 """Display-only projection: never transfer credentials or execute saved tools."""
 turns=[];current=None
 for message in session.get('messages',[]):
  if message.get('operational_snapshot') or message.get('tool_name'):continue
  if message.get('shell_result'):
   if current is not None:current['shell']=True;current['responses'].append(message.get('text',''))
   continue
  if message.get('role')=='user':
   current={'number':len(turns)+1,'prompt':message.get('text',''),'responses':[],'calls':0}
   turns.append(current)
  elif message.get('role')=='assistant' and current is not None:
   current['calls']+=sum(item.get('type')=='function_call' for item in message.get('provider_items',[]))
   if message.get('text'):current['responses'].append(message['text'])
 return turns[-3:]

class Foreground:
 def __enter__(self):
  self.process=None
  if sys.stdin.isatty() and sys.stdout.isatty() and os.getenv('LCA_REMOTE_PLAIN')!='1' and os.getenv('TERM')!='dumb':
   self.temp=tempfile.TemporaryDirectory(prefix='lca-foreground-')
   path=Path(self.temp.name)/'commands.jsonl';path.touch(mode=0o600)
   self.commands=path.open()
   read_fd,write_fd=os.pipe()
   self.events=os.fdopen(write_fd,'w')
   try:
    self.process=subprocess.Popen(['lua5.5',str(Path(__file__).with_suffix('.lua')),str(path),str(read_fd)],pass_fds=(read_fd,))
   except BaseException:
    self.events.close();self.commands.close();self.temp.cleanup();raise
   finally:os.close(read_fd)
  return self
 def event(self,event):
  if self.process:
   if self.process.poll() is not None:return
   try:self.events.write(json.dumps(event)+'\n');self.events.flush()
   except BrokenPipeError:pass
  elif event['type'] in ('output','notice'):
   print(event['text'],end='' if event['type']=='output' else '\n',flush=True)
 def notice(self,text):self.event({'type':'notice','text':text})
 def output(self,text):self.event({'type':'output','text':text})
 def state(self,state):
  backup=state.get('checkpoint',{})
  if backup.get('saved_at') and backup['saved_at']!=getattr(self,'last_backup',None):
   self.notice('Remote checkpoint saved '+time.strftime('%H:%M:%S',time.localtime(backup['saved_at'])))
   self.last_backup=backup['saved_at']
  if state.get('checkpoint_error') and state['checkpoint_error']!=getattr(self,'backup_error',None):
   self.notice('Backup failed; further work paused. '+state['checkpoint_error'])
  self.backup_error=state.get('checkpoint_error')
  stream=getattr(self,'live_stream',None)
  if not stream or stream['finished']:self.history(state)
  self.event({'type':'state','phase':state['phase'],'active':state.get('active'),
              'live_events':bool(state.get('capabilities',{}).get('live_events') and state.get('live') and not state['live'].get('error'))})
 def history(self,state):
  if self.process and isinstance(state.get('session'),dict):
   turns=recent_turns(state['session'])
   if turns!=getattr(self,'last_turns',None):
    self.event({'type':'history','turns':turns,'session_id':state['session'].get('id','lca')})
    self.last_turns=turns
 def live(self,state,read):
  if not self.process or not state.get('capabilities',{}).get('live_events'):return
  initial=not getattr(self,'live_initialized',False)
  self.live_initialized=True
  meta=state.get('live')
  if not isinstance(meta,dict) or not re.fullmatch(r'[A-Za-z0-9-]+',meta.get('id','')):return
  stream=getattr(self,'live_stream',None)
  def drain(stream):
   chunk=read('/tmp/lca-handoff/events/'+stream['id']+'.jsonl',stream['offset'])
   stream['offset']+=len(chunk);stream['buffer']+=chunk
   while b'\n' in stream['buffer']:
    line,stream['buffer']=stream['buffer'].split(b'\n',1)
    # Guest display previews are byte-bounded and may end inside a UTF-8 glyph.
    item=json.loads(line.decode('utf-8',errors='replace'))
    if item['id']!=stream['id']:raise RuntimeError('Remote event journal identity changed')
    self.event({'type':'live','event':item})
    if item['kind']=='finish':stream['finished']=True
   if len(stream['buffer'])>262144:raise RuntimeError('Remote display event exceeds size limit')
   return bool(chunk)
  if stream and stream['id']!=meta['id'] and not stream['finished']:
   try:drain(stream)
   except (RuntimeError,ValueError):
    self.event({'type':'live_reset'});self.notice('Older live activity is unavailable; restoring completed history.');stream['finished']=True
   if not stream['finished']:return
  if not stream or stream['id']!=meta['id']:
   if initial:self.history(state)
   stream={'id':meta['id'],'offset':0,'buffer':b'','finished':initial and bool(meta.get('complete'))}
   self.live_stream=stream
  if not stream['finished']:
   try:more=drain(stream)
   except (RuntimeError,ValueError):
    self.event({'type':'live_reset'});self.notice('Live activity is unavailable; completed history will still appear.');stream['finished']=True;return
   if not more and meta.get('complete') and not stream['finished']:
    self.event({'type':'live_reset'});self.notice('Live display interrupted; restoring completed history.');stream['finished']=True
 def readline(self,timeout=1):
  if not self.process:
   return sys.stdin.readline() if select.select([sys.stdin],[],[],timeout)[0] else None
  until=time.monotonic()+timeout
  while True:
   offset=self.commands.tell();line=self.commands.readline()
   if line and not line.endswith('\n'):self.commands.seek(offset);line=''
   if line:return json.loads(line)['text']+'\n'
   if self.process.poll() is not None:
    if self.process.returncode:raise RuntimeError('Local remote-session display failed; session remains remote')
    return ''
   if time.monotonic()>=until:return None
   time.sleep(.05)
 def __exit__(self,*args):
  if self.process:
   try:
    try:self.events.close()
    except BrokenPipeError:pass
    self.process.wait(timeout=5)
   finally:
    if self.process.poll() is None:self.process.terminate();self.process.wait()
    self.commands.close();self.temp.cleanup()

File: microvm/foreground.lua
-- Host-only view. This process never runs the remote session or its tools.
local uv=require('luv')
local json=require('cjson')
local tui=require('agent.tui')
local M={}
local function commit_visible_reply(app,text)
 local protocol=require('agent.tool_protocol')
 local visible=protocol.strip_tool_results(protocol.strip_tool_calls(text))
 if visible:match('%S') then app:commit_assistant(visible) end
end
local function live(app,event)
 local item=event.event
 app.live_sequences=app.live_sequences or {}
 local previous=app.live_sequences[item.id] or 0
 if item.seq<=previous then return end
 app.live_sequences[item.id]=item.seq
 if item.kind=='begin' then
  app.remote_live=true;app.remote_history=true;app.remote_seen=app.remote_seen or {}
  app.live_id=item.id;app.live_filter=tui.StreamFilter.new();app.work_update_ids={};app.live_degraded=false
  app.busy=true;app.session_badge='remote · working';app:auto_advance_effect()
  app.state.turn_sequence=item.number-1;app.state:submit(item.prompt);app.river_seed=item.session_id
  app:commit_user(item.prompt)
  app.live_history_key=item.prompt:sub(1,1)~='/' and (string.format('%d',item.number)..':'..item.prompt) or nil
 elseif item.id~=app.live_id then return
 elseif item.kind=='token' then
  local visible,activity=app.live_filter:feed(item.text)
  if activity then app.state:model_activity(activity) end
  if visible~='' then app.state:model_stream(visible) end
 elseif item.kind=='tool' then app.state:tool_event(item.event)
 elseif item.kind=='thinking' then app.state:reviewing(item.info);app.live_filter=tui.StreamFilter.new()
 elseif item.kind=='activity' then
  if item.activity.type=='assistant_commentary' then app:commit_commentary(item.activity.item,'live_commentary') end
  app.state:model_activity(item.activity)
 elseif item.kind=='response' then app:commit_work_update(item.response)
 elseif item.kind=='answer' then
  local protocol=require('agent.tool_protocol')
  local text=protocol.strip_tool_results(protocol.strip_tool_calls(app:final_response_text(item.response)))
  local usage=item.usage or {};local tokens=tonumber(usage.prompt_tokens)
  local cache=usage.cache_available and tokens and tokens>0 and (tonumber(usage.cached_tokens) or 0)/tokens*100 or nil
  app.state:assistant_complete(text,{tokens=tokens,cache_percent=cache});app:commit_river();commit_visible_reply(app,text)
 elseif item.kind=='command_output' then
  local lines={};for line in (item.text..'\n'):gmatch('(.-)\n') do lines[#lines+1]='shell › '..line end
  app:commit_lines(lines)
 elseif item.kind=='gap' then app.live_degraded=true;app:commit_lines({item.text})
 elseif item.kind=='finish' then
  local tail=app.live_filter:finish();if tail~='' then app.state:model_stream(tail) end
  app:commit_river(item.error and 'failed' or nil)
  if app.live_history_key and not app.live_degraded then app.remote_seen[app.live_history_key]=true end
  app.remote_live=false;app.busy=false;app.session_badge='remote · ready'
  if item.error then app.state:notice(item.error,'error');app.state.mode='failed';app:commit_lines({'error › '..item.error})
  else app.state:listen() end
 end
end
function M.apply(app,event)
 if event.type=='live_reset' then app.remote_live=false;app.busy=false;app.live_id=nil;return end
 if event.type=='live' then live(app,event);return end
 if event.type=='history' then
  app.remote_history=true
  app.remote_seen=app.remote_seen or {}
  for _,turn in ipairs(event.turns) do
   local key=string.format('%d',turn.number)..':'..turn.prompt
   -- state.json contains completed-turn snapshots; polling must not replay them.
   if not app.remote_seen[key] and #turn.responses>0 then
    app.remote_seen[key]=true
    app:commit_user(turn.prompt)
    if turn.shell then
     local lines={};for line in (table.concat(turn.responses,'\n')..'\n'):gmatch('(.-)\n') do lines[#lines+1]='shell › '..line end
     app:commit_lines(lines)
    else
    for i=1,#turn.responses-1 do
     commit_visible_reply(app,turn.responses[i])
    end
    if turn.calls>0 then
     local lines=require('agent.river_divider').render({width=select(1,app:_size()),
      color=app.backend:supports_color(),seed=event.session_id,turn=turn.number})
     lines[#lines+1]=string.format('%d tool calls · restored history',turn.calls)
     app:commit_lines(lines)
    end
    commit_visible_reply(app,turn.responses[#turn.responses])
    end
   end
  end
 elseif event.type=='output' then
  if app.live_transport then return end
  app.remote_tail=(app.remote_tail or '')..event.text
  local lines={}
  while app.remote_tail:find('\n',1,true) do
   local line,rest=app.remote_tail:match('^(.-)\n(.*)$')
   if app.remote_history then
    if line:match('^> ') then app.remote_raw_command=line:match('^> /')~=nil end
    if app.remote_raw_command then lines[#lines+1]=line end
   else lines[#lines+1]=line end
   app.remote_tail=rest
  end
  if #lines>0 then app:commit_lines(lines) end
 elseif event.type=='notice' then
  app.state:notice(event.text)
  app:commit_lines({event.text})
 elseif event.type=='state' then
  app.live_transport=event.live_events==true
  if app.live_transport and (event.phase=='running' or app.remote_live and event.phase=='idle') then return end
  local labels={checkpointing='backup paused',running='working',idle='ready',ready='ready',suspending='sleeping',waking='waking',reconnecting='reconnecting',failed='failed',stopped='stopped'}
  app.session_badge='remote · '..(labels[event.phase] or event.phase)
  local key=event.phase..':'..(type(event.active)=='string' and event.active or '')
  if key==app.remote_key then return end
  app.remote_key=key
  app.busy=event.phase=='running'
  if app.busy then
   app:auto_advance_effect();app.state:submit('Remote LCA')
   app.state.model_phase='working in MicroVM'
  else
   app.state:listen()
   app.state:notice(event.phase=='suspending' and 'MicroVM sleeping · send a prompt to wake' or 'Remote · '..event.phase)
  end
 end
end
function M.run(command_path,event_fd)
 local stage=tostring(os.getenv('LCA_TOOL_STAGE') or ''):lower()
 local app=tui.App.new({effect=os.getenv('LCA_TUI_EFFECT'),tool_stage=not ({['0']=true,['false']=true,off=true,no=true})[stage or '']})
 app.session_badge='remote · connecting'
 app.input_hint='Enter sends · Ctrl-C/D detaches · /local brings home'
 local commands=assert(io.open(command_path,'a'));commands:setvbuf('no')
 local function send(text) commands:write(json.encode({text=text})..'\n') end
 -- Ctrl-C detaches, never claims to cancel a remote tool.
 app._handle_action=function(self,action)
  if action and action.type=='cancel' then self.exit_requested=true
  else tui.App._handle_action(self,action) end
 end
 local input=uv.new_pipe(false);input:open(assert(tonumber(event_fd)))
 local buffer=''
 local ok,err=tui.with_terminal(app.terminal,app.renderer,function()
  app.state:notice('Connecting to remote session…')
  app:start_io()
  input:read_start(function(read_err,chunk)
   if read_err then app.fatal_error=tostring(read_err);app.exit_requested=true;return end
   if not chunk then app.exit_requested=true;return end
   buffer=buffer..chunk
   while buffer:find('\n',1,true) do
    local line,rest=buffer:match('^(.-)\n(.*)$');buffer=rest
    local decoded,event=pcall(json.decode,line)
    if decoded then M.apply(app,event) else app.fatal_error='invalid remote display event';app.exit_requested=true end
   end
  end)
  while not app.exit_requested do
   local line=app:next_submission()
   if line then
    local effect=line:match('^/effect%s+(%S+)%s*$')
    if effect then
     local changed,why=pcall(app.set_effect,app,effect)
     if not changed then app.state:notice(tostring(why),'error') end
    elseif line=='/tools on' then app:set_tool_stage(true)
    elseif line=='/tools off' then app:set_tool_stage(false)
    elseif line=='/tools' then app.state:notice('tool stage · '..(app.tool_stage and 'on' or 'off')..' · /tools on|off')
    elseif line=='/river' then app:commit_river_details()
    elseif line=='/bg' or line=='/background' then send('/detach');break
    elseif line=='/cloud' then app.state:notice('Already attached to the cloud session.')
    elseif line=='/detach' or line=='/local' then send(line);break
    elseif line:sub(1,1)=='/' and not ({test=true,status=true,context=true,reasoning=true,['service-tier']=true})[line:match('^/(%S+)')] then
     app.state:notice('Remote commands: /test, /status, /context, /reasoning, /service-tier, /effect, /tools, /river, /local, /bg, /detach','error')
    else send(line);app.state:notice('Submitting input…') end
   end
  end
  if app.exit_requested then send('/detach') end
  if app.fatal_error then error(app.fatal_error) end
  return true
 end)
 app:stop_io();input:read_stop();input:close();commands:close();uv.run('nowait')
 if not ok then error(err) end
end
if arg and arg[0] and arg[0]:match('/foreground%.lua$') then M.run(assert(arg[1]),assert(arg[2])) end
return M
